src/validation.cpp

```diff
@@ -2486,25 +2486,15 @@ class ConnectTrace {
     CTxMemPool &pool;
 
 public:
-    explicit ConnectTrace(CTxMemPool &_pool) : blocksConnected(1), pool(_pool) {}
+    explicit ConnectTrace(CTxMemPool &_pool) : pool(_pool) {}
 
     void BlockConnected(CBlockIndex* pindex, std::shared_ptr<const CBlock> pblock) {
-        assert(!blocksConnected.back().pindex);
         assert(pindex);
         assert(pblock);
-        blocksConnected.back().pindex = pindex;
-        blocksConnected.back().pblock = std::move(pblock);
-        blocksConnected.emplace_back();
+        blocksConnected.emplace_back(pindex, std::move(pblock));
     }
 
     std::vector<PerBlockConnectTrace>& GetBlocksConnected() {
-        // We always keep one extra block at the end of our list because
-        // blocks are added after all the conflicted transactions have
-        // been filled in. Thus, the last entry should always be an empty
-        // one waiting for the transactions from the next block. We pop
-        // the last entry here to make sure the list we return is sane.
-        assert(!blocksConnected.back().pindex);
-        blocksConnected.pop_back();
```

Comment:
I prefer to leave this as a separate commit since I think it's easier to review removing the logic piece-by-piece. If other reviewers agree with you, then I'm happy to squash the commits.

Reply:
I think @ariard's comment was about more than commit management: 
The comment starting in line 2499 is referring to conflicted transactions and seems outdated but isn't adjusted in any commit - leading to the question if the logic of always having an empty entry at the back of the vector and popping it before returning `blocksConnected` in `GetBlocksConnected()` still makes any sense or could be removed/simplified as well?